File: packages/ts/src/parsers.ts
/**
 * Manifest parsers. Each parser extracts the set of dependency identifiers used for framework and
 * testing fingerprinting. Identifiers are normalized per-language so the Python twin can match them.
 */
import { readFileSync } from "node:fs";
import { join } from "node:path";
import TOML from "@iarna/toml";
import { XMLParser } from "fast-xml-parser";
import { basename } from "./signals.js";

export interface DependencyPools {
  /** JavaScript/TypeScript package names (from every package.json). */
  js: Set<string>;
  /** Python distribution names, PEP 503-ish normalized (lowercase). */
  py: Set<string>;
  /** JVM coordinates: both `group:artifact` and bare `artifact`. */
  java: Set<string>;
  /** Go module paths. */
  go: Set<string>;
  /** Rust crate names. */
  rust: Set<string>;
}

function read(root: string, rel: string): string {
  return readFileSync(join(root, rel), "utf8");
}

function normalizePy(name: string): string {
  return name
    .trim()
    .toLowerCase()
    .replace(/[-_.]+/g, "-");
}

/** Strip a PEP 508 requirement down to its distribution name. */
function pep508Name(spec: string): string | null {
  const s = spec.trim();
  if (!s || s.startsWith("#")) return null;
  const m = s.match(/^([A-Za-z0-9._-]+)/);
  return m ? normalizePy(m[1]!) : null;
}

function addPackageJson(root: string, rel: string, pools: DependencyPools): void {
  try {
    const pkg = JSON.parse(read(root, rel)) as Record<string, unknown>;
    for (const field of [
      "dependencies",
      "devDependencies",
      "peerDependencies",
      "optionalDependencies",
    ]) {
      const deps = pkg[field];
      if (deps && typeof deps === "object") {
        for (const name of Object.keys(deps as object)) pools.js.add(name);
      }
    }
  } catch {
    /* ignore malformed manifest */
  }
}

function addPyproject(root: string, rel: string, pools: DependencyPools): void {
  try {
    const doc = TOML.parse(read(root, rel)) as any;
    const projDeps = doc?.project?.dependencies;
    if (Array.isArray(projDeps)) {
      for (const spec of projDeps) {
        const n = pep508Name(String(spec));
        if (n) pools.py.add(n);
      }
    }
    const optional = doc?.project?.["optional-dependencies"];
    if (optional && typeof optional === "object") {
      for (const arr of Object.values(optional)) {
        if (Array.isArray(arr))
          for (const spec of arr) {
            const n = pep508Name(String(spec));
            if (n) pools.py.add(n);
          }
      }
    }
    const poetry = doc?.tool?.poetry;
    if (poetry?.dependencies && typeof poetry.dependencies === "object") {
      for (const name of Object.keys(poetry.dependencies)) {
        if (name.toLowerCase() !== "python") pools.py.add(normalizePy(name));
      }
    }
    if (poetry?.group && typeof poetry.group === "object") {
      for (const g of Object.values<any>(poetry.group)) {
        if (g?.dependencies && typeof g.dependencies === "object") {
          for (const name of Object.keys(g.dependencies)) {
            if (name.toLowerCase() !== "python") pools.py.add(normalizePy(name));
          }
        }
      }
    }
  } catch {
    /* ignore */
  }
}

function addRequirements(root: string, rel: string, pools: DependencyPools): void {
  try {
    for (const line of read(root, rel).split(/\r?\n/)) {
      const n = pep508Name(line);
      if (n) pools.py.add(n);
    }
  } catch {
    /* ignore */
  }
}

function addSetupPy(root: string, rel: string, pools: DependencyPools): void {
  try {
    const src = read(root, rel);
    const m = src.match(/install_requires\s*=\s*\[([^\]]*)\]/s);
    if (m) {
      for (const q of m[1]!.match(/["']([^"']+)["']/g) ?? []) {
        const n = pep508Name(q.replace(/["']/g, ""));
        if (n) pools.py.add(n);
      }
    }
  } catch {
    /* ignore */
  }
}

function addPom(root: string, rel: string, pools: DependencyPools): void {
  try {
    const parser = new XMLParser({ ignoreAttributes: true, isArray: () => false });
    const doc = parser.parse(read(root, rel)) as any;
    let deps = doc?.project?.dependencies?.dependency;
    if (!deps) return;
    if (!Array.isArray(deps)) deps = [deps];
    for (const d of deps) {
      const group = d?.groupId != null ? String(d.groupId).trim() : "";
      const artifact = d?.artifactId != null ? String(d.artifactId).trim() : "";
      if (group && artifact) pools.java.add(`${group}:${artifact}`);
      if (artifact) pools.java.add(artifact);
    }
  } catch {
    /* ignore */
  }
}

function addGradle(root: string, rel: string, pools: DependencyPools): void {
  try {
    const src = read(root, rel);
    for (const m of src.matchAll(/["']([\w.-]+):([\w.-]+):[\w.$-]+["']/g)) {
      pools.java.add(`${m[1]}:${m[2]}`);
      pools.java.add(m[2]!);
    }
  } catch {
    /* ignore */
  }
}

function addGoMod(root: string, rel: string, pools: DependencyPools): void {
  try {
    const src = read(root, rel);
    // require ( ... ) blocks and single-line requires
    for (const m of src.matchAll(/^\s*(?:require\s+)?([a-zA-Z0-9._~-]+(?:\/[a-zA-Z0-9._~-]+)+)\s+v/gm)) {
      pools.go.add(m[1]!);
    }
  } catch {
    /* ignore */
  }
}

function addCargo(root: string, rel: string, pools: DependencyPools): void {
  try {
    const doc = TOML.parse(read(root, rel)) as any;
    for (const section of ["dependencies", "dev-dependencies", "build-dependencies"]) {
      const tbl = doc?.[section];
      if (tbl && typeof tbl === "object") for (const name of Object.keys(tbl)) pools.rust.add(name);
    }
  } catch {
    /* ignore */
  }
}

/** Parse every recognized manifest under `root`, returning per-language dependency pools. */
export function parseManifests(root: string, files: string[]): DependencyPools {
  const pools: DependencyPools = {
    js: new Set(),
    py: new Set(),
    java: new Set(),
    go: new Set(),
    rust: new Set(),
  };
  for (const rel of files) {
    const base = basename(rel);
    if (base === "package.json") addPackageJson(root, rel, pools);
    else if (base === "pyproject.toml") addPyproject(root, rel, pools);
    else if (base === "requirements.txt") addRequirements(root, rel, pools);
    else if (base === "setup.py") addSetupPy(root, rel, pools);
    else if (base === "pom.xml") addPom(root, rel, pools);
    else if (base === "build.gradle" || base === "build.gradle.kts") addGradle(root, rel, pools);
    else if (base === "go.mod") addGoMod(root, rel, pools);
    else if (base === "Cargo.toml") addCargo(root, rel, pools);
  }
  return pools;
}

File: packages/py/src/repo_fingerprint/parsers.py
"""Manifest parsers producing per-language dependency pools for fingerprinting.

Identifier normalization matches the TypeScript twin so the two agree on framework detection.
"""
from __future__ import annotations

import json
import re
import tomllib
import xml.etree.ElementTree as ET
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from .signals import basename


@dataclass
class DependencyPools:
    js: set[str] = field(default_factory=set)
    py: set[str] = field(default_factory=set)
    java: set[str] = field(default_factory=set)
    go: set[str] = field(default_factory=set)
    rust: set[str] = field(default_factory=set)


def _read(root: str, rel: str) -> str:
    return (Path(root) / rel).read_text(encoding="utf-8")


def _normalize_py(name: str) -> str:
    return re.sub(r"[-_.]+", "-", name.strip().lower())


def _pep508_name(spec: str) -> str | None:
    s = spec.strip()
    if not s or s.startswith("#"):
        return None
    m = re.match(r"^([A-Za-z0-9._-]+)", s)
    return _normalize_py(m.group(1)) if m else None


def _local(tag: str) -> str:
    return tag.rsplit("}", 1)[-1]


def _add_package_json(root: str, rel: str, pools: DependencyPools) -> None:
    try:
        pkg = json.loads(_read(root, rel))
    except Exception:
        return
    for field_name in ("dependencies", "devDependencies", "peerDependencies", "optionalDependencies"):
        deps = pkg.get(field_name)
        if isinstance(deps, dict):
            pools.js.update(deps.keys())


def _add_pyproject(root: str, rel: str, pools: DependencyPools) -> None:
    try:
        doc = tomllib.loads(_read(root, rel))
    except Exception:
        return
    project = doc.get("project", {})
    for spec in project.get("dependencies", []) or []:
        n = _pep508_name(str(spec))
        if n:
            pools.py.add(n)
    for arr in (project.get("optional-dependencies", {}) or {}).values():
        for spec in arr or []:
            n = _pep508_name(str(spec))
            if n:
                pools.py.add(n)
    poetry = doc.get("tool", {}).get("poetry", {})
    for name in (poetry.get("dependencies", {}) or {}):
        if name.lower() != "python":
            pools.py.add(_normalize_py(name))
    for group in (poetry.get("group", {}) or {}).values():
        for name in (group.get("dependencies", {}) or {}):
            if name.lower() != "python":
                pools.py.add(_normalize_py(name))


def _add_requirements(root: str, rel: str, pools: DependencyPools) -> None:
    try:
        text = _read(root, rel)
    except Exception:
        return
    for line in text.splitlines():
        n = _pep508_name(line)
        if n:
            pools.py.add(n)


def _add_setup_py(root: str, rel: str, pools: DependencyPools) -> None:
    try:
        src = _read(root, rel)
    except Exception:
        return
    m = re.search(r"install_requires\s*=\s*\[(.*?)\]", src, re.DOTALL)
    if m:
        for q in re.findall(r"""["']([^"']+)["']""", m.group(1)):
            n = _pep508_name(q)
            if n:
                pools.py.add(n)


def _add_pom(root: str, rel: str, pools: DependencyPools) -> None:
    try:
        tree = ET.parse(str(Path(root) / rel))
    except Exception:
        return
    for dep in tree.iter():
        if _local(dep.tag) != "dependency":
            continue
        group = ""
        artifact = ""
        for child in dep:
            ln = _local(child.tag)
            if ln == "groupId":
                group = (child.text or "").strip()
            elif ln == "artifactId":
                artifact = (child.text or "").strip()
        if group and artifact:
            pools.java.add(f"{group}:{artifact}")
        if artifact:
            pools.java.add(artifact)


def _add_gradle(root: str, rel: str, pools: DependencyPools) -> None:
    try:
        src = _read(root, rel)
    except Exception:
        return
    for m in re.finditer(r"""["']([\w.-]+):([\w.-]+):[\w.$-]+["']""", src):
        pools.java.add(f"{m.group(1)}:{m.group(2)}")
        pools.java.add(m.group(2))


def _add_go_mod(root: str, rel: str, pools: DependencyPools) -> None:
    try:
        src = _read(root, rel)
    except Exception:
        return
    for m in re.finditer(
        r"^\s*(?:require\s+)?([a-zA-Z0-9._~-]+(?:/[a-zA-Z0-9._~-]+)+)\s+v", src, re.MULTILINE
    ):
        pools.go.add(m.group(1))


def _add_cargo(root: str, rel: str, pools: DependencyPools) -> None:
    try:
        doc = tomllib.loads(_read(root, rel))
    except Exception:
        return
    for section in ("dependencies", "dev-dependencies", "build-dependencies"):
        tbl = doc.get(section)
        if isinstance(tbl, dict):
            pools.rust.update(tbl.keys())


def parse_manifests(root: str, files: list[str]) -> DependencyPools:
    pools = DependencyPools()
    for rel in files:
        base = basename(rel)
        if base == "package.json":
            _add_package_json(root, rel, pools)
        elif base == "pyproject.toml":
            _add_pyproject(root, rel, pools)
        elif base == "requirements.txt":
            _add_requirements(root, rel, pools)
        elif base == "setup.py":
            _add_setup_py(root, rel, pools)
        elif base == "pom.xml":
            _add_pom(root, rel, pools)
        elif base in ("build.gradle", "build.gradle.kts"):
            _add_gradle(root, rel, pools)
        elif base == "go.mod":
            _add_go_mod(root, rel, pools)
        elif base == "Cargo.toml":
            _add_cargo(root, rel, pools)
    return pools
